### PR TITLE
Implement special viewport binding for amp-inabox

### PR DESCRIPTION
**Pull request by lannka**
For #5700

### CODE REVIEW

File: src/inabox/inabox-viewport.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+  /** @override */
+  connect() {
+    this.iframeClient_.makeRequest(
+        MessageType.SEND_POSITIONS, MessageType.POSITION,
+        data => {
+          dev().fine(TAG, 'Position changed: ', data);
+          const oldViewportRect = this.viewportRect_;
+          const oldSelfRect = this.selfRect_;
+          this.viewportRect_ = data.viewport;
+          this.selfRect_ = data.target;
```
Comment (jridgewell) [inline src/inabox/inabox-viewport.js]: `target`?

File: src/inabox/inabox-viewport.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+    // Before hearing from host doc, we're blind about the viewport position
+    // and iframe position. Positive width/height ensure a render start.
+    /** @private {!../layout-rect.LayoutRectDef} */
+    this.viewportRect_ = layoutRectLtwh(0, 0, 10, 10);
+
+    // To not trigger amp-analytics visibility immediately,
+    // we assume we're right below the fold.
+    /** @private {!../layout-rect.LayoutRectDef} */
+    this.selfRect_ = layoutRectLtwh(0, 10, 10, 10);
```
Comment (jridgewell) [inline src/inabox/inabox-viewport.js]: What is `selfRect`?

File: src/inabox/inabox-viewport.js [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const {!Window} */
+    this.win = win;
+
+    /** @private @const {!Observable} */
+    this.scrollObservable_ = new Observable();
```
Comment (zhouyx) [inline src/inabox/inabox-viewport.js]: I think @chenshay had a pr that lazily create observable(). Do we also want to do that for inabox host?

File: src/inabox/inabox-viewport.js (hunk 1 above)
Comment (zhouyx) [inline src/inabox/inabox-viewport.js]: https://github.com/ampproject/amphtml/blob/master/ads/inabox/position-observer.js#L90

File: src/inabox/inabox-viewport.js [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+          dev().fine(TAG, 'Position changed: ', data);
+          const oldViewportRect = this.viewportRect_;
+          const oldSelfRect = this.selfRect_;
+          this.viewportRect_ = data.viewport;
+          this.selfRect_ = data.target;
+          if (isMoved(this.selfRect_, oldSelfRect)) {
+            // Remeasure all AMP elements once iframe position is changed.
+            // Because all layout boxes are calculated relatively to the
+            // iframe position.
+            this.remeasureAllElements_();
```
Comment (zhouyx) [inline src/inabox/inabox-viewport.js]: I am not sure about this. When an amp element is inabox. its `layoutBox` value should be relative to the box. Changing the position of the box shouldn't affect their cached layoutBox value?

File: src/inabox/inabox-viewport.js (hunk 1 above)
Comment (jridgewell) [inline src/inabox/inabox-viewport.js]: Ok, now why "target"? This is supposed to be the `<iframe>`'s rect from the host doc, right?

Comment (lannka) [inline src/inabox/inabox-viewport.js]: Added more docs. Hope it's clearer now.

File: src/inabox/inabox-viewport.js (hunk 3 above)
Comment (lannka) [inline src/inabox/inabox-viewport.js]: I actually now think it does not worth it for many cases. For example, here it's guaranteed to be created once viewport get initialized.

Comment (zhouyx) [inline src/inabox/inabox-viewport.js]: but I assume inabox-viewport won't be scrollable. Then why would we waste setting this up in constructor?

Comment (lannka) [inline src/inabox/inabox-viewport.js]: The scrolling events is from host doc via post message. Resources, analytics, many other things are still listening to viewport for scroll changes.

**Comment by lannka** [comment]
Tests added. It's ready for review now. PTAL

File: src/inabox/inabox-viewport.js (hunk 3 above)
Comment (zhouyx) [inline src/inabox/inabox-viewport.js]: Ahhh! scrolling of the host doc.

File: src/inabox/inabox-viewport.js (hunk 4 above)
Comment (zhouyx) [inline src/inabox/inabox-viewport.js]: ping here~

Comment (lannka) [inline src/inabox/inabox-viewport.js]: we make all element positions relative to the document, that's how we get the whole resource scheduling work

File: src/inabox/inabox-viewport.js [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+  connect() {
+    this.iframeClient_.makeRequest(
+        MessageType.SEND_POSITIONS, MessageType.POSITION,
+        data => {
+          dev().fine(TAG, 'Position changed: ', data);
+          const oldViewportRect = this.viewportRect_;
+          const oldSelfRect = this.boxRect_;
+          this.viewportRect_ = data.viewport;
+          this.boxRect_ = data.target;
+          if (isMoved(this.boxRect_, oldSelfRect)) {
```
Comment (zhouyx) [inline src/inabox/inabox-viewport.js]: I can't find the logic where we remeasure elements when window size changes. If not I still think we need to call `#remeasureAllElements_` in the case of `isResized(this.boxRect_)`.

File: src/inabox/inabox-viewport.js [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+  return /** @type {!Viewport} */(getServiceForDoc(
+      ampdoc, 'viewport', () => viewport));
+}
+
+/**
+ * @param {!Window} win
+ * @returns {string}
+ */
+function getRandom(win) {
+  return String(win.Math.random()).substr(2);
```
Comment (lannka) [inline src/inabox/inabox-viewport.js]: 2 is not the length, but the starting index

File: src/inabox/inabox-viewport.js (hunk 5 above)
Comment (zhouyx) [inline src/inabox/inabox-viewport.js]: As we discussed offline, you wanna add a comment here?

File: src/inabox/inabox-viewport.js (hunk 6 above)
Comment (zhouyx) [inline src/inabox/inabox-viewport.js]: Ah!

File: src/inabox/inabox-viewport.js (hunk 5 above)
Comment (lannka) [inline src/inabox/inabox-viewport.js]: Added `isResized` check as well. And added a TODO for firing a DOM mutation event when we handle them.

**Comment by lannka** [comment]
I removed `inabox-viewer.js` for now. We can do the optimization later time once the viewer-impl is more stabilized. 

PTAL

File: src/inabox/inabox-viewport.js [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+          this.viewportRect_ = data.viewport;
+          this.boxRect_ = data.target;
+          if (isChanged(this.boxRect_, oldSelfRect)) {
+            // Remeasure all AMP elements once iframe position is changed.
+            // Because all layout boxes are calculated relatively to the
+            // iframe position.
+            this.remeasureAllElements_();
+            // TODO: fire DOM mutation event once we handle them
+          }
+          if (isResized(this.viewportRect_, oldViewportRect)) {
```
Comment (jridgewell) [inline src/inabox/inabox-viewport.js]: Don't we need to remeasure here, too?

Comment (zhouyx) [inline src/inabox/inabox-viewport.js]: this's viewport change. The element pos to the doc should not be affected i believe

File: 3p/iframe-messaging-client.js [hunk 8]
```diff
@@ -103,11 +103,19 @@ export class IframeMessagingClient {
     });
   }
 
+  /**
+   * @param {!Window} win
+   */
   setHostWindow(win) {
+    /** @private {!Window} */
```
Comment (lannka) [inline 3p/iframe-messaging-client.js]: I'm wondering if this is for code style or it makes a difference to the compiler?

Comment (jridgewell) [inline 3p/iframe-messaging-client.js]: You've mutated the hidden class here, which deoptimizes property access on the object. You've got to set each property in the constructor, even if it's to `null`.

Comment (lannka) [inline 3p/iframe-messaging-client.js]: done

File: 3p/iframe-messaging-client.js [hunk 9]
```diff
@@ -103,11 +103,19 @@ export class IframeMessagingClient {
...
     this.hostWindow_ = win;
   }
 
+  /**
+   * @param {string} sentinel
+   */
   setSentinel(sentinel) {
+    /** @private {string} */
```
Comment (lannka) [inline 3p/iframe-messaging-client.js]: done

**Comment by lannka** [comment]
Merging now. Will address further comments in separate PR.

**Pull request merged**